<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>NEXTFLEX Combo-Seq · NEXTFLEX Combo-Seq Documentation</title><script data-outdated-warner src="assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href>NEXTFLEX Combo-Seq Documentation</a></span></div><form class="docs-search" action="search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>NEXTFLEX Combo-Seq</a><ul class="internal"><li><a class="tocitem" href="#Index"><span>Index</span></a></li><li><a class="tocitem" href="#Functions"><span>Functions</span></a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>NEXTFLEX Combo-Seq</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>NEXTFLEX Combo-Seq</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/cookienocreams/combo_seq/blob/main/docs/src/index.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="NEXTFLEX-Combo-Seq"><a class="docs-heading-anchor" href="#NEXTFLEX-Combo-Seq">NEXTFLEX Combo-Seq</a><a id="NEXTFLEX-Combo-Seq-1"></a><a class="docs-heading-anchor-permalink" href="#NEXTFLEX-Combo-Seq" title="Permalink"></a></h1><h2 id="Index"><a class="docs-heading-anchor" href="#Index">Index</a><a id="Index-1"></a><a class="docs-heading-anchor-permalink" href="#Index" title="Permalink"></a></h2><ul><li><a href="#combo_seq.Config"><code>combo_seq.Config</code></a></li><li><a href="#combo_seq.MissingReferenceError"><code>combo_seq.MissingReferenceError</code></a></li><li><a href="#combo_seq.align_with_salmon"><code>combo_seq.align_with_salmon</code></a></li><li><a href="#combo_seq.calculate_read_length_distribution"><code>combo_seq.calculate_read_length_distribution</code></a></li><li><a href="#combo_seq.calculate_salmon_metrics"><code>combo_seq.calculate_salmon_metrics</code></a></li><li><a href="#combo_seq.capture_target_files"><code>combo_seq.capture_target_files</code></a></li><li><a href="#combo_seq.create_reference_file"><code>combo_seq.create_reference_file</code></a></li><li><a href="#combo_seq.create_target_organism_fasta_file"><code>combo_seq.create_target_organism_fasta_file</code></a></li><li><a href="#combo_seq.find_common_rnas"><code>combo_seq.find_common_rnas</code></a></li><li><a href="#combo_seq.generate_mirna_counts"><code>combo_seq.generate_mirna_counts</code></a></li><li><a href="#combo_seq.get_genus_names"><code>combo_seq.get_genus_names</code></a></li><li><a href="#combo_seq.get_read_q_score!"><code>combo_seq.get_read_q_score!</code></a></li><li><a href="#combo_seq.make_bowtie2_reference"><code>combo_seq.make_bowtie2_reference</code></a></li><li><a href="#combo_seq.make_metrics_violin_plot"><code>combo_seq.make_metrics_violin_plot</code></a></li><li><a href="#combo_seq.make_salmon_reference"><code>combo_seq.make_salmon_reference</code></a></li><li><a href="#combo_seq.mirna_discovery_calculation"><code>combo_seq.mirna_discovery_calculation</code></a></li><li><a href="#combo_seq.parse_fastq_files"><code>combo_seq.parse_fastq_files</code></a></li><li><a href="#combo_seq.plot_clustering"><code>combo_seq.plot_clustering</code></a></li><li><a href="#combo_seq.plot_fragment_lengths"><code>combo_seq.plot_fragment_lengths</code></a></li><li><a href="#combo_seq.plot_metrics"><code>combo_seq.plot_metrics</code></a></li><li><a href="#combo_seq.plot_mirna_counts"><code>combo_seq.plot_mirna_counts</code></a></li><li><a href="#combo_seq.progress_bar_update"><code>combo_seq.progress_bar_update</code></a></li><li><a href="#combo_seq.remove_files"><code>combo_seq.remove_files</code></a></li><li><a href="#combo_seq.remove_intermediate_files"><code>combo_seq.remove_intermediate_files</code></a></li><li><a href="#combo_seq.trim_adapters"><code>combo_seq.trim_adapters</code></a></li><li><a href="#combo_seq.write_common_rna_file"><code>combo_seq.write_common_rna_file</code></a></li></ul><h2 id="Functions"><a class="docs-heading-anchor" href="#Functions">Functions</a><a id="Functions-1"></a><a class="docs-heading-anchor-permalink" href="#Functions" title="Permalink"></a></h2><p>Documentation for combo_seq.jl</p><article class="docstring"><header><a class="docstring-binding" id="combo_seq.MissingReferenceError" href="#combo_seq.MissingReferenceError"><code>combo_seq.MissingReferenceError</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">struct MissingReferenceError
    message::String
end</code></pre><p>A structure that represents a custom exception for a missing reference.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/combo_seq/blob/d1fb2e5bc522cbf908c75c332de03d1ebc058782/src/combo_seq.jl#L59-L65">source</a></section></article><div class="admonition is-warning"><header class="admonition-header">Missing docstring.</header><div class="admonition-body"><p>Missing docstring for <code>IncorrectOrganismError</code>. Check Documenter&#39;s build log for details.</p></div></div><div class="admonition is-warning"><header class="admonition-header">Missing docstring.</header><div class="admonition-body"><p>Missing docstring for <code>MissingFastqsError</code>. Check Documenter&#39;s build log for details.</p></div></div><article class="docstring"><header><a class="docstring-binding" id="combo_seq.Config" href="#combo_seq.Config"><code>combo_seq.Config</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">struct Config
    need_reference::Bool
    mrna::Union{String, Nothing}
    fasta::Union{String, Nothing}
    transcript::Union{String, Nothing}
    genome::Union{String, Nothing}
    organism::Union{String, Nothing}
    threads::Int
end</code></pre><p>A structure that stores the command-line arguments.</p><p><strong>Functionality</strong></p><ul><li>Contains a check to ensure that a Salmon and miRNA fasta file are passed as arguments. </li></ul><p>If they were not, an error will be returned. This is to reduce the chances of an alignment  errors downstream.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/combo_seq/blob/d1fb2e5bc522cbf908c75c332de03d1ebc058782/src/combo_seq.jl#L81-L99">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="combo_seq.get_genus_names" href="#combo_seq.get_genus_names"><code>combo_seq.get_genus_names</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Create a file with a list of unique species names given a miRNA mature or hairpin fasta.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/combo_seq/blob/d1fb2e5bc522cbf908c75c332de03d1ebc058782/src/combo_seq.jl#L965-L967">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="combo_seq.create_target_organism_fasta_file" href="#combo_seq.create_target_organism_fasta_file"><code>combo_seq.create_target_organism_fasta_file</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Function takes in mirBase mature miRNA fasta file with data from all available organisms and pulls out only the miRNA data pertaining to the target organism.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/combo_seq/blob/d1fb2e5bc522cbf908c75c332de03d1ebc058782/src/combo_seq.jl#L983-L986">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="combo_seq.capture_target_files" href="#combo_seq.capture_target_files"><code>combo_seq.capture_target_files</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">capture_target_files(files_to_capture::AbstractString, directory::AbstractString=&quot;.&quot;)</code></pre><p>List all files in a directory. </p><p>Check to see if each file contains the target file&#39;s string. </p><p><strong>Example</strong></p><pre><code class="language-julia hljs">julia&gt; capture_target_files(&quot;.txt&quot;)
3-element Vector{String}:
 &quot;file1.txt&quot;
 &quot;file2.txt&quot;
 &quot;file3.txt&quot;</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/combo_seq/blob/d1fb2e5bc522cbf908c75c332de03d1ebc058782/src/combo_seq.jl#L149-L164">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="combo_seq.progress_bar_update" href="#combo_seq.progress_bar_update"><code>combo_seq.progress_bar_update</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Update progress bar on the command line.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/combo_seq/blob/d1fb2e5bc522cbf908c75c332de03d1ebc058782/src/combo_seq.jl#L169-L171">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="combo_seq.get_read_q_score!" href="#combo_seq.get_read_q_score!"><code>combo_seq.get_read_q_score!</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">get_read_q_score!(line::String, q_score_list::Vector{Number})</code></pre><p>Calculate the quality score for a given quality string.</p><p>This function accepts ASCII-encoded quality score and produces the average q score  using a scalar value that estimates the &#39;total error probability&#39; for that record. It means that if you want to calculate average quality score, you don&#39;t just sum up all the phred scores and find the average, but rather consider the  probability distribution of the scores.</p><p>The Phred score for a base <code>Q</code> is calculated as <code>Q = ASCII value of quality score - 33</code>. The error probability <code>P</code> for that base is then calculated as <code>P = 10^(-Q/10)</code>. Then these probabilities are summed up for all the bases to derive total error.</p><p><strong>Arguments</strong></p><p><code>q_score</code> - The quality scores of the current line. <code>q_score_list</code> - The vector containing all calculated quality scores.</p><p><strong>Returns</strong></p><p>The updated q score list the line&#39;s average quality score.</p><p><strong>Reference</strong></p><p>Illumina&#39;s explanation on quality scores and their corresponding error rates: <a href="https://www.illumina.com/science/technology/next-generation-sequencing/plan-experiments/quality-scores.html">https://www.illumina.com/science/technology/next-generation-sequencing/plan-experiments/quality-scores.html</a></p><p><strong>Example</strong></p><pre><code class="language-julia hljs">julia&gt; get_read_q_score!(&quot;FGFEEE&lt;FC&lt;FGGGGGGGGGGGFFGFG8&lt;@8CFFGF8EFGGCGFGGGGGGFG&quot;, [36.2, 35.9])
3-element Vector{Float64}:
 36.2
 35.9
 35.7</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/combo_seq/blob/d1fb2e5bc522cbf908c75c332de03d1ebc058782/src/combo_seq.jl#L186-L223">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="combo_seq.make_bowtie2_reference" href="#combo_seq.make_bowtie2_reference"><code>combo_seq.make_bowtie2_reference</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Make miRNA bowtie2 reference to align fastqs with.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/combo_seq/blob/d1fb2e5bc522cbf908c75c332de03d1ebc058782/src/combo_seq.jl#L243-L245">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="combo_seq.make_salmon_reference" href="#combo_seq.make_salmon_reference"><code>combo_seq.make_salmon_reference</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Make Salmon reference for mRNA alignment.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/combo_seq/blob/d1fb2e5bc522cbf908c75c332de03d1ebc058782/src/combo_seq.jl#L256-L258">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="combo_seq.create_reference_file" href="#combo_seq.create_reference_file"><code>combo_seq.create_reference_file</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Function asks the user if they would like to download a non-human reference for use by Salmon during the alignment step. If yes, the files are downloaded and the genome names are extracted to name the Salmon directory.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/combo_seq/blob/d1fb2e5bc522cbf908c75c332de03d1ebc058782/src/combo_seq.jl#L300-L304">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="combo_seq.parse_fastq_files" href="#combo_seq.parse_fastq_files"><code>combo_seq.parse_fastq_files</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">parse_fastq_files(fastqs::Vector{String}, sample_names::Vector{SubString{String}})</code></pre><p>Calculate read counts, percent dimers, and average Q scores for a given set of fastq files.</p><p><strong>Parameters</strong></p><ul><li><code>fastqs</code>: A vector of filenames for the fastq files.</li><li><code>sample_names</code>: A vector of sample names corresponding to the fastq files.</li></ul><p><strong>Returns</strong></p><p>Three dictionaries containing read counts, percent dimers, and average Q scores  indexed by sample name.</p><p><strong>Example</strong></p><pre><code class="language-julia hljs">julia&gt; parse_fastq_files([&quot;sample1.fastq.gz&quot;, &quot;sample2.fastq.gz&quot;, &quot;sample3.fastq.gz&quot;], 
                        [&quot;sample1&quot;, &quot;sample2&quot;, &quot;sample3&quot;])
# Returns:
# - A dictionary of read counts.
# - A dictionary of percent dimers.
# - A dictionary of average Q scores.</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/combo_seq/blob/d1fb2e5bc522cbf908c75c332de03d1ebc058782/src/combo_seq.jl#L406-L428">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="combo_seq.trim_adapters" href="#combo_seq.trim_adapters"><code>combo_seq.trim_adapters</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">trim_adapters(fastqs::Vector{String}, sample_names::Vector{SubString{String}})</code></pre><p>Trim the 3&#39; adapter from each read in the provided fastq files.</p><p><strong>Combo-Seq Adapters:</strong></p><p>5&#39; Adapter: GATCGTCGGACTGTAGAACTCTGAACNNNN </p><p>3&#39; Adapter: AAAAAAAAAA</p><p>3&#39; bases with a quality score below 20 are trimmed. Reads shorter than 16 bases or  those that weren&#39;t trimmed are discarded.</p><p><strong>Parameters</strong></p><ul><li><code>fastqs</code>: A vector of filenames for the fastq files.</li><li><code>sample_names</code>: A vector of sample names corresponding to the fastq files.</li></ul><p><strong>Returns</strong></p><p>A list of filenames for the trimmed fastq files.</p><p><strong>Example</strong></p><pre><code class="language-julia hljs">julia&gt; trim_adapters([&quot;sample1.fastq.gz&quot;, &quot;sample2.fastq.gz&quot;, &quot;sample3.fastq.gz&quot;], 
                    [&quot;sample1&quot;, &quot;sample2&quot;, &quot;sample3&quot;])
# Returns:
[&quot;sample1.cut.fastq&quot;, &quot;sample2.cut.fastq&quot;, &quot;sample3.cut.fastq&quot;]</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/combo_seq/blob/d1fb2e5bc522cbf908c75c332de03d1ebc058782/src/combo_seq.jl#L484-L511">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="combo_seq.align_with_salmon" href="#combo_seq.align_with_salmon"><code>combo_seq.align_with_salmon</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">align_with_salmon(trimmed_fastq_files::Vector{String}
                    , salmon_reference::AbstractString
                    , sample_names::Vector{SubString{String}}
                    , need_reference::Bool
                    )</code></pre><p>This function aligns trimmed fASTQ files to a reference genome using Salmon.</p><p>The library type (libType) is set to stranded forward (SF) since the Combo-Seq protocol is stranded in that direction.</p><p><strong>Arguments</strong></p><ul><li><code>trimmed_fastq_files</code>: A vector of strings containing the paths to the trimmed fASTQ files.</li><li><code>salmon_reference</code>: A string containing the path to the Salmon reference genome.</li><li><code>sample_names</code>: A vector of strings containing the names of the samples.</li><li><code>need_reference</code>: A string indicating whether or not a custom Salmon index was created.</li></ul><p><strong>Returns</strong></p><p>The function returns a vector of strings containing the paths to the Salmon output files.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/combo_seq/blob/d1fb2e5bc522cbf908c75c332de03d1ebc058782/src/combo_seq.jl#L541-L562">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="combo_seq.calculate_salmon_metrics" href="#combo_seq.calculate_salmon_metrics"><code>combo_seq.calculate_salmon_metrics</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">calculate_salmon_metrics(trimmed_fastq_files::Vector{String}
                        , read_count_dict::Dict{String, Int64}
                        , sample_names::Vector{SubString{String}}
                        , dimer_count_dict::Dict{String, Float64}
                        , q_score_dict::Dict{String, Number}
                        )</code></pre><p>Align fastq with specified reference RNA types.</p><p>Divide reads aligned with total reads to calculate percent alignment. Store  alignment information in a dictionary.</p><p>#Example</p><pre><code class="language-julia hljs">julia&gt; calculate_salmon_metrics(&quot;sample1.cut.fastq&quot;
                        , Dict(&quot;sample1&quot; =&gt; 6390314, &quot;sample2&quot; =&gt; 5000000, &quot;sample3&quot; =&gt; 7052928))
# Returns:
# - &quot;ComboSeq_metrics.csv&quot;</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/combo_seq/blob/d1fb2e5bc522cbf908c75c332de03d1ebc058782/src/combo_seq.jl#L728-L748">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="combo_seq.generate_mirna_counts" href="#combo_seq.generate_mirna_counts"><code>combo_seq.generate_mirna_counts</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">generate_mirna_counts(input_sam_file::AbstractString)</code></pre><p>This function creates a DataFrame containing all aligned miRNA and the number of times  they appear.</p><p><strong>Returns</strong></p><p>A DataFrame with the following columns:</p><ul><li><code>name</code>: The name of the miRNA.</li><li><code>count</code>: The number of times the miRNA appears in the SAM file.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/combo_seq/blob/d1fb2e5bc522cbf908c75c332de03d1ebc058782/src/combo_seq.jl#L1048-L1060">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="combo_seq.make_metrics_violin_plot" href="#combo_seq.make_metrics_violin_plot"><code>combo_seq.make_metrics_violin_plot</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Create violin plot of each sample&#39;s RNA aligment metrics</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/combo_seq/blob/d1fb2e5bc522cbf908c75c332de03d1ebc058782/src/combo_seq.jl#L846-L848">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="combo_seq.mirna_discovery_calculation" href="#combo_seq.mirna_discovery_calculation"><code>combo_seq.mirna_discovery_calculation</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">mirna_discovery_calculation(trimmed_fastq_files::Vector{String}
                            , sample_names::Vector{SubString{String}}
                            , bowtie2_reference_name::AbstractString
                            , organism_name::AbstractString
                            , need_reference::AbstractString
                            )</code></pre><p>Align trimmed fastq files to the single organism miRNA bowtie2 reference. </p><p>The bowtie2 output SAM is input into function to generate miRNA read counts.</p><p>#Example</p><pre><code class="language-julia hljs">julia&gt; mirna_discovery_calculation([&quot;sample1.cut.fastq&quot;,&quot;sample2.cut.fastq&quot;,&quot;sample3.cut.fastq&quot;]
                                , [&quot;sample1&quot;, &quot;sample2&quot;, &quot;sample3&quot;])
3-element Vector{DataFrame}:
469x2 DataFrame
 Row │ name             count
     │ String           Int64
─────┼────────────────────────
   1 │ hsa-miR-1307-3p      1
   2 │ hsa-miR-425-5p      58
   3 │ hsa-miR-488-5p       2
 [...]

 3-element Vector{String}:
 &quot;sample1.miRNA.sam&quot;
 &quot;sample2.miRNA.sam&quot;
 &quot;sample3.miRNA.sam&quot;</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/combo_seq/blob/d1fb2e5bc522cbf908c75c332de03d1ebc058782/src/combo_seq.jl#L1097-L1128">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="combo_seq.plot_mirna_counts" href="#combo_seq.plot_mirna_counts"><code>combo_seq.plot_mirna_counts</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">plot_mirna_counts(mirna_counts_dfs::Vector{DataFrame}
                    , sample_names::Vector{SubString{String}}
                    )</code></pre><p>Plot counts of the number of unique miRNA each sample aligned to.</p><p>#Example</p><pre><code class="language-julia hljs">julia&gt; plot_mirna_counts([sample1_counts_df,sample2_counts_df,sample3_counts_df]
                        , [&quot;sample1&quot;, &quot;sample2&quot;, &quot;sample3&quot;])
3-element Vector{String}:
 &quot;sample1_miRNA_counts.csv&quot;
 &quot;sample2_miRNA_counts.csv&quot;
 &quot;sample3_miRNA_counts.csv&quot;</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/combo_seq/blob/d1fb2e5bc522cbf908c75c332de03d1ebc058782/src/combo_seq.jl#L1167-L1183">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="combo_seq.calculate_read_length_distribution" href="#combo_seq.calculate_read_length_distribution"><code>combo_seq.calculate_read_length_distribution</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Use aligned SAM files to create a read length distribution file.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/combo_seq/blob/d1fb2e5bc522cbf908c75c332de03d1ebc058782/src/combo_seq.jl#L623-L625">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="combo_seq.plot_fragment_lengths" href="#combo_seq.plot_fragment_lengths"><code>combo_seq.plot_fragment_lengths</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Create barplot of each sample&#39;s fragment lengths based on the number of reads found at each length</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/combo_seq/blob/d1fb2e5bc522cbf908c75c332de03d1ebc058782/src/combo_seq.jl#L677-L679">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="combo_seq.plot_metrics" href="#combo_seq.plot_metrics"><code>combo_seq.plot_metrics</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Create barplot of each sample&#39;s RNA aligment metrics.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/combo_seq/blob/d1fb2e5bc522cbf908c75c332de03d1ebc058782/src/combo_seq.jl#L901-L903">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="combo_seq.find_common_rnas" href="#combo_seq.find_common_rnas"><code>combo_seq.find_common_rnas</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">find_common_rnas(counts_files::Vector{String}
                , sample_names::Vector{SubString{String}} 
                ; rna_type::String=&quot;miRNA&quot;, read_count_dict::Dict{String
                , Int64}=Dict{String, Int64}()
                )</code></pre><p>Return the RNAs present in all samples, together with their counts and reads per million reads (RPM) or transcripts per million (TPM).</p><p><strong>Arguments</strong></p><ul><li><code>counts_files</code>: A vector of files containing sample RNA counts.</li><li><code>sample_names</code>: A vector of sample names.</li><li><code>rna_type</code>: A string specifying the type of RNA (&quot;miRNA&quot; or &quot;mRNA&quot;). Default is &quot;miRNA&quot;.</li><li><code>read_count_dict</code>: (Optional for &quot;mRNA&quot; type) A dictionary where keys are sample names and </li></ul><p>values are the total number of reads  for the corresponding sample. This is required for &quot;miRNA&quot; type to compute RPM values.</p><p><strong>Returns</strong></p><ul><li>A Tuple of three items:<ol><li>A list of RNAs that are common to all samples.</li><li>A list of dictionaries where each dictionary contains the RNA counts for one sample.</li><li>A list of dictionaries where each dictionary contains the RPM (for miRNA) or TPM </li></ol>(for mRNA) values for the RNAs in one sample.</li></ul><p><strong>Example</strong></p><p>For miRNA:</p><pre><code class="language-julia hljs">julia&gt; find_common_rnas([&quot;sample1_miRNA_counts.csv&quot;, &quot;sample2_miRNA_counts.csv&quot;],
                        [&quot;sample1&quot;, &quot;sample2&quot;],
                        rna_type=&quot;miRNA&quot;,
                        read_count_dict=Dict(&quot;sample1&quot; =&gt; 6390314, &quot;sample2&quot; =&gt; 5000000))</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/combo_seq/blob/d1fb2e5bc522cbf908c75c332de03d1ebc058782/src/combo_seq.jl#L1350-L1383">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="combo_seq.write_common_rna_file" href="#combo_seq.write_common_rna_file"><code>combo_seq.write_common_rna_file</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">write_common_rna_file(rna_names::Base.KeySet{String, Dict{String, Int64}},
                      rna_info::Vector{Dict{String, Number}},
                      normalization_info::Vector{Dict{String, Number}},
                      sample_names::Vector{SubString{String}};
                      rna_type::String=&quot;miRNA&quot;)</code></pre><p>Write the common RNAs found in all samples to an output file. This function generates two sets of files for each sample:  one containing raw RNA counts and another containing normalized values (RPM for miRNA or TPM for mRNA). It then  combines these files for all samples to produce comprehensive outputs, and subsequently conducts a principal component  analysis and UMAP on the common RNA data.</p><p><strong>Arguments</strong></p><ul><li><code>rna_names</code>: A set containing the names of RNAs that are common across all samples.</li><li><code>rna_info</code>: A vector of dictionaries with each dictionary containing the RNA counts for one sample.</li><li><code>normalization_info</code>: A vector of dictionaries, each containing the RPM (for miRNA) or TPM (for mRNA) values for the RNAs in one sample.</li><li><code>sample_names</code>: A vector of sample names.</li><li><code>rna_type</code>: (Optional) A string specifying the type of RNA (&quot;miRNA&quot; or &quot;mRNA&quot;). Default is &quot;miRNA&quot;.</li></ul><p><strong>Outputs</strong></p><ul><li>Generates files for each sample named like <code>sample_common_miRNAs.tsv</code> and <code>sample_common_miRNAs_RPM.tsv</code> (or their mRNA equivalents).</li><li>Combines these files to produce comprehensive outputs named <code>Common_miRNAs.tsv</code> and <code>Common_RPM_miRNAs.tsv</code> (or their mRNA equivalents).</li><li>Conducts principal component analysis and UMAP based on common RNA data.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/combo_seq/blob/d1fb2e5bc522cbf908c75c332de03d1ebc058782/src/combo_seq.jl#L1435-L1458">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="combo_seq.plot_clustering" href="#combo_seq.plot_clustering"><code>combo_seq.plot_clustering</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">plot_clustering(Common_miRNA_File::String)</code></pre><p>Generates Principal Component Analysis (PCA) and Uniform Manifold Approximation and  Projection (UMAP) plots for libraries given a common RNA file.</p><p>This function performs the following steps:</p><ol><li>Reads the common RNA file and creates a DataFrame.</li><li>Extracts sample names and [mi/m]RNA names.</li><li>Ensures that there are at least two samples and more than one miRNA present in the dataset.</li><li>Transforms the data using PCA and prepares it for plotting.</li><li>If there are at least two principal components, plots the PCA and saves it as &quot;common<em>[mi/m]RNA</em>PCA.png&quot;.</li><li>If there are at least three principal components, performs UMAP dimensionality reduction and clustering using K-medoids.</li><li>Plots the UMAP and saves it as &quot;common<em>[mi/m]RNA</em>UMAP.png&quot;.</li><li>Writes the PCA and UMAP information to separate CSV files for cluster tracking.</li></ol><p><strong>Arguments</strong></p><ul><li><code>Common_[mi/m]RNA_File::String</code>: Path to the common RNA RPM/TPM file.</li></ul><p><strong>Outputs</strong></p><ul><li>Creates and saves PCA and UMAP plots as &quot;common<em>[mi/m]RNA</em>PCA.png&quot; and &quot;common<em>[mi/m]RNA</em>UMAP.png&quot;.</li><li>Writes PCA and UMAP information to &quot;PCA<em>information.csv&quot; and &quot;UMAP</em>information.csv&quot;.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/combo_seq/blob/d1fb2e5bc522cbf908c75c332de03d1ebc058782/src/combo_seq.jl#L1507-L1529">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="combo_seq.remove_files" href="#combo_seq.remove_files"><code>combo_seq.remove_files</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Spot remove unecessary intermediate files.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/combo_seq/blob/d1fb2e5bc522cbf908c75c332de03d1ebc058782/src/combo_seq.jl#L1611-L1613">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="combo_seq.remove_intermediate_files" href="#combo_seq.remove_intermediate_files"><code>combo_seq.remove_intermediate_files</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Remove all intermediate files.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/combo_seq/blob/d1fb2e5bc522cbf908c75c332de03d1ebc058782/src/combo_seq.jl#L1624-L1626">source</a></section></article></article><nav class="docs-footer"><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.25 on <span class="colophon-date" title="Tuesday 9 January 2024 22:54">Tuesday 9 January 2024</span>. Using Julia version 1.10.0-beta2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
